src/lib.rs

```
@@ -490,22 +576,32 @@ where
     }
 }
 
-async fn handle_tcp_connection<S>(addr: SocketAddr, receiver: &mut S)
+async fn handle_tcp_connection<F, R, S, I>(addr: SocketAddr, mut f: F, receiver: &mut S)
 where
     S: Stream<Item = Result<Bytes, std::io::Error>>,
     S: Unpin,
+    I: tokio::io::AsyncRead + tokio::io::AsyncWrite + Send + Unpin,
+    F: FnMut(TcpStream) -> R + Send + Sync + Clone + 'static,
+    R: Future<Output = Result<I, std::io::Error>> + Send,
 {
-    // Try connect
-    let mut tcp_stream = match TcpStream::connect(addr).await {
+    let try_connect = async move {
```

Why is this in an `async` block?

---

I believe it's to enable the use of `?`, which is used twice in this block. I refactored this in 68961da so the function propagates its errors (I did the same in `handle_udp_connection` for consistency). The callers of these functions still ignore the error, but that's probably a more logical place to handle/log it if someone wants to make another PR.